plugins/kinetic-dp/fu-kinetic-dp-secure-aux-isp.c

```diff
@@ -0,0 +1,873 @@
+/*
+ * Copyright (C) 2021 Jeffrey Lin <jlin@kinet-ic.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "fu-kinetic-dp-secure-aux-isp.h"
+
+#include "fu-kinetic-dp-aux-isp.h"
+#include "fu-kinetic-dp-connection.h"
+#include "fu-kinetic-dp-aux-dpcd.h"
+
+/* OUI of MegaChips America */
+#define MCA_OUI_BYTE_0				0x00
+#define MCA_OUI_BYTE_1				0x60
+#define MCA_OUI_BYTE_2				0xAD
+
+/* Kinetic proprietary DPCD fields for Jaguar/Mustang, for both application and ISP driver */
+#define DPCD_ADDR_FLOAT_CMD_STATUS_REG		0x0050D
+#define DPCD_ADDR_FLOAT_PARAM_REG		0x0050E
+
+/* Below DPCD registers are used while running application */
+#define DPCD_ADDR_FLOAT_CUSTOMER_FW_MIN_REV	0x00514
+#define DPCD_SIZE_FLOAT_CUSTOMER_FW_MIN_REV	1
+#define DPCD_ADDR_FLOAT_CUSTOMER_PROJ_ID	0x00515
+#define DPCD_SIZE_FLOAT_CUSTOMER_PROJ_ID	1
+#define DPCD_ADDR_FLOAT_PRODUCT_TYPE		0x00516
+#define DPCD_SIZE_FLOAT_PRODUCT_TYPE		1
+
+/* Below DPCD registers are used while running ISP driver */
+#define DPCD_ADDR_FLOAT_ISP_REPLY_LEN_REG	0x00513
+#define DPCD_SIZE_FLOAT_ISP_REPLY_LEN_REG	1		/* 0x00513			*/
+
+#define DPCD_ADDR_FLOAT_ISP_REPLY_DATA_REG	0x00514		/* While running ISP driver	*/
+#define DPCD_SIZE_FLOAT_ISP_REPLY_DATA_REG	12		/* 0x00514 ~ 0x0051F		*/
+
+#define DPCD_ADDR_KT_AUX_WIN			0x80000ul
+#define DPCD_SIZE_KT_AUX_WIN			0x8000ul	/* 0x80000ul ~ 0x87FFF, 32 KB	*/
+#define DPCD_ADDR_KT_AUX_WIN_END		(DPCD_ADDR_KT_AUX_WIN +  DPCD_SIZE_KT_AUX_WIN - 1)
+
+#define INIT_CRC16				0x1021
+
+/* Each device should have a corresponding ISP info instance */
+static guint32 isp_payload_procd_size;
+static guint32 isp_procd_size;
+static guint32 isp_total_data_size;
+
+static guint16 read_flash_prog_time;
+static guint16 flash_id;
+static guint16 flash_size;
+
+static gboolean is_isp_secure_auth_mode;
+
+static guint16
+_gen_crc16 (guint16 accum, guint8 data_in)
```

Comment:
remove this and use fu_common_crc16 please

Reply:
iIn our implementation, I think CRC16 is calculated with CRC-16-CCITT polynomial.
Is fu_common_crc16() based on CRC-16-IBM polynomial?
I'm not very sure if these two are different CRC-16 algorithms.

Reply:
This should help you https://github.com/fwupd/fwupd/pull/3452 -- feel free to cherry pick that commit into your branch.

Reply:
These days I've studied for CRC-16 things, it's more complicated than what I could imagine.
I tried to find some reference codes to compare the outputs among Kinetic's CRC-16 function, fu_common_crc16(), and other reference implementations(CRC-16, CRC16-CCITT).
I ported the CCITT function from a library called LibCRC.
But I found that the output of fu_common_crc16(init value: 0xFFFF, polynomial: 0x12) is not the same as the output of libcrc.
My references are as below links:
http://srecord.sourceforge.net/crc16-ccitt.html
https://reveng.sourceforge.io/crc-catalogue/16.htm
[LibCRC]
https://www.lammertbies.nl/comm/info/crc-calculation
https://www.libcrc.org/

According to one of above articles, there're many incorrect CRC-16 implementations on the web.
And there're so many variants(parameters) to implement.
May I know how do you implement fu_common_crc16() function and what's your reference?

Back to the PR, I think our CRC-16 implementation just works as a checksum mechanism while transmitting the payload.
I don't think it's a standard CRC16-CCITT implementation(I don't know the history either...).
I can't just replace it with fu_common_crc16() because of the backward-compatible issue in Windows tools released to our customers.
So my first thought is to rename our CRC16 implementation as a proprietary CRC-16 "static" function.
Would it be acceptable?

Thanks a lot.

Reply:
> So my first thought is to rename our CRC16 implementation as a proprietary CRC-16 "static" function

Yes, that's probably the best plan. If you can add some documentation explaining that I'd appreciate it.